govuk_chat_evaluation/rag_answers/evaluate.py

```diff
@@ -1,31 +1,225 @@
+    # calculate aggregated results and exports results to CSV files
+    aggregation.export_to_csvs(_output_dir, prefix="alessia_test")
+
+    print("Evaluation Results:")
+    print(aggregation.summary)
+
+
+class EvaluationResults:
+    def __init__(self, evaluation_config: EvaluationConfig, cases: list[LLMTestCase]):
+        self.evaluation_config = evaluation_config
+        self.metrics = evaluation_config.get_metric_instances()
+        self.llm_judge = evaluation_config.llm_judge_instance
+        self.n_runs = evaluation_config.n_runs
+        self.cases = cases
+        self.evaluation_results : list[TestResult] = []
+        
+    def run_deepeval_evaluation(self, **kwargs) -> None:
+        """"
+        Run the Deepval evaluation on the given models and metrics
+
+        Args:
+            cases : List of test cases to evaluate
+            metrics : List of metrics to use for evaluation
+            n_runs : Number of runs to perform for the evaluation
+            **kwargs: Additional arguments to pass to the deepeval.evaluation function
+
+        Returns:
+            List of evaluation results for each model
+
+        """
+
+        with Timer("Running Deepval Evaluation"):
+            logger.info("Running Deepval evaluation")
+
+            all_evaluation_runs = []
+
+            for i in range(self.n_runs):
+                logger.info(f"Running evaluation iteration {i+1}/{self.n_runs}...")
+
+                evaluation_run = deepeval_evaluate(
+                    test_cases=cast(list[LLMTestCase | MLLMTestCase], self.cases), # hint to type checker: this is a List[LLMTestCase | MLLMTestCase], even though it's just List[LLMTestCase]
+                    metrics=self.metrics, # type: ignore
+                    **kwargs,  # pass additional arguments dynamically 
+                )
+
+                all_evaluation_runs.append(evaluation_run.test_results)  # Store results per run
+
+        logger.info("Deepval evaluation complete")
+        
+        self.evaluation_results = [results for run in all_evaluation_runs for results in run]
+
+        return None
+
+
+    def get_results(self):
+        """Return the computed evaluation results."""
+        return self.evaluation_results
+    
+    @staticmethod
+    def _serialise_eval_result(result: TestResult) -> dict:
+        result_dict = asdict(result)
+        if "metrics_data" in result_dict:
+            result_dict["metrics_data"] = [
+                metric.model_dump() if isinstance(metric, BaseModel) else metric
+                for metric in result_dict["metrics_data"]
+            ]
+        return result_dict
+    
+    def to_serialised_dicts(self) -> list[dict[str, Any]]:
+        return [self._serialise_eval_result(result) for result in self.evaluation_results]
+
+
+
+class AggregatedResults:
+    def __init__(
+        self,
+        evaluation_results: list,
+        additional_metric_fields: list[str] = [],
+        additional_result_fields: list[str] = []
```

Comment:
Shall we remove these until they are being used?

I found them a bit confusing myself and looked for any usage. I did wonder if rather than being additional they're more options (or even better would be that this class didn't need to know about them and could derive them from the results it's given).

Reply:
Happy to remove them for now. They have only been needed for the `NonContradiction` metric but you are right, worth re-adding them in case only then in case.